examples/src/dining_cryptographers.rs

```
@@ -0,0 +1,325 @@
+        RequiredMessageParts, RequiredMessages, Round, RoundId, TransitionInfo,
+    },
+    session::SessionParameters,
+    signature::{self, Keypair},
+};
+use rand_core::{CryptoRngCore, OsRng};
+use serde::{Deserialize, Serialize};
+use tracing::{debug, info};
+
+#[derive(Debug)]
+pub struct DiningCryptographersProtocol;
+#[derive(displaydoc::Display, Debug, Clone, Serialize, Deserialize)]
+/// bla
+pub enum DiningError {
+    /// bla
+    R1Err,
+    /// bla
+    R2Err,
+}
+// TODO: for protocols that do not have "provable errors" we should have a shortcut and/or blanket impls.
+// TODO: Why does ProtocolError require Display?
+impl<Id> ProtocolError<Id> for DiningError {
+    type AssociatedData = ();
+    fn required_messages(&self) -> RequiredMessages {
+        match self {
+            DiningError::R1Err => RequiredMessages::new(RequiredMessageParts::direct_message(), None, None),
+            DiningError::R2Err => RequiredMessages::new(RequiredMessageParts::direct_message(), None, None),
+        }
+    }
+
+    fn verify_messages_constitute_error(
+        &self,
+        format: &BoxedFormat,
+        guilty_party: &Id,
+        shared_randomness: &[u8],
+        associated_data: &Self::AssociatedData,
+        message: ProtocolMessage,
+        previous_messages: BTreeMap<RoundId, ProtocolMessage>,
+        combined_echos: BTreeMap<RoundId, BTreeMap<Id, EchoBroadcast>>,
+    ) -> Result<(), ProtocolValidationError> {
+        match self {
+            DiningError::R1Err => Ok(()),
+            DiningError::R2Err => Ok(()),
+        }
+    }
+}
+
+impl<Id> Protocol<Id> for DiningCryptographersProtocol {
+    // TODO: Is this the right place for the final outcome of the protocol?
+    // XOR/¬XOR of the two bits of the diners (one their own cointoss, the other shared with their neighbour).
+    type Result = bool;
+
+    type ProtocolError = DiningError;
+
+    // TODO: Having blanket impls for these methods would cut down on the boilerplate a bit.
```

True, but what would be the default behavior? If it's just ignoring the error, there is a danger of forgetting to implement them. 

That said, I am not happy with these methods. They are far away from the message type, and not very convenient to write and debug. There should be a better way - somehow bundle these checks with the message type itself.

---

You mean: protocols are so different that there's no reasonable default here? That's fair I guess. OTOH there's a limit to the amnount of hand-holding a library can provide beyond "make it easy to do the right thing". Tricky to know, but I guess the it's an indication that writing examples is useful to see where the sore spots are.